src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfT.cs

```diff
@@ -138,6 +138,15 @@ public override bool GetMemberAndWriteJson(object obj, ref WriteStack state, Utf
         {
             T value = Get!(obj);
 
+            if (Options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.IgnoreCycle &&
+                !Converter.IsValueType && value != null &&
+                state.ReferenceResolver.ContainsReferenceForCycleDetection(value))
+            {
+                // If a reference cycle is detected, treat value as null.
+                value = default!;
```

Comment:
Are we potentially overwriting user data here (mutating properties/fields on objects passed when serializing)?

Reply:
I don't think so, `value` is an independent variable declared in this method. I will add an assertion that guards for this anyway.